### PR TITLE
Suppress Compiler Warnings (kernel)

### PR DESCRIPTION
**Pull request by satorg**
This is an ongoing effort for suppressing compiler warnings that keep plaguing the project.

Suppresses the warnings in kernel\* modules for all Scala versions.

### CODE REVIEW

File: core/src/main/scala/cats/TraverseFilter.scala
Comment (satorg) [inline core/src/main/scala/cats/TraverseFilter.scala:135]: Raises the "parameter `H: Hash[A]` not used" warning because it is really so.

I'd prefer to keep this warning in place until Cats receive a suitable `HashSet` implementation (perhaps by merging #4185).

**Review by johnynek — APPROVED** [review]
Thanks for the cleanup!

**Comment by satorg** [comment]
Seems there can be some issues with migrating to Scala 3.1.x versions (see discussion in https://github.com/http4s/http4s/pull/6300#pullrequestreview-943891503)

Perhaps, it may make sense to make use of `val _ = someUnusedParam` pattern in this PR as well...

**Comment by armanbilge** [comment]
The other option is we continue to keep fatal warnings disabled on Scala 3 😕

**Comment by satorg** [comment]
I don't think it's going to help... Scala 3.0 silently ignores unused parameters along with any kind of `@nowarn` annotation. Whereas Scala 3.1 will be failing on `@nowarn("cat=unused-params")` anyway regardless of the fatal warnings option.

**Comment by satorg** [comment]
Added more thoughts on how to overcome the compatibility issue: https://github.com/http4s/http4s/pull/6300#discussion_r851824642

I'm considering to take that approach, but sincerely open to discussion :)

File: core/src/main/scala/cats/data/NonEmptySet.scala
Comment (satorg) [inline core/src/main/scala/cats/data/NonEmptySet.scala:432]: Can someone explain (or guess at least) why does this change not concern Mima? I removed the unused implicit `Order` parameter but seems that Mima is completely fine about that.

File: core/src/main/scala/cats/data/NonEmptySet.scala
Comment (satorg) [inline core/src/main/scala/cats/data/NonEmptySet.scala:431]: Can someone explain (or guess at least) why does this change not concern Mima? I removed the unused implicit `Order` parameter but seems that Mima is completely fine about that.

Comment (satorg) [inline core/src/main/scala/cats/data/NonEmptySet.scala:431]: Can someone explain (or guess at least) why does this change not concern Mima? I removed the unused implicit `Order` parameter but seems that Mima is completely fine about that.

File: kernel/src/main/scala-2.12/cats/kernel/compat/unused.scala
Comment (satorg) [inline kernel/src/main/scala-2.12/cats/kernel/compat/unused.scala:31]: Just to draw more attention to this Scala version compatibility trick.

File: core/src/main/scala/cats/data/NonEmptySet.scala
Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Um, that seems suspect. Let me look into it, hopefully there's a good explanation ...

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Ok, I'm really confused ... but is this all private??

`NonEmptySetInstances0` is private, and extended by `NonEmptySetInstances`, which is private, which is implemented by `NonEmptySetImpl`, also private.

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Right.. But the implicit method itself is public and should be publicly accessible as a part of the `NonEmptySet` object API.

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: That's the thing .. AFAICT there _is_ no (public) `object NonEmptySet` anywhere in Cats 🙃

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: That's correct... It is just a val: `val NonEmptySet = NonEmptySetImpt`...

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: ..and the same trick is used for `NonEmptyMap` and `NonEmptyChain`

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Ah, thanks, I'm sorry. Didn't realize that's how it's implemented.
https://github.com/typelevel/cats/blob/708cbaba2e6ac07a5ac1819b6c18b0c83a2faaba/core/src/main/scala/cats/data/package.scala#L39

The bad news is, we won't get any MiMa support on this :( The reason is that the type of `NonEmptySet` is `NonEmptySetImpl.type` aka `NonEmptySetImpl$`, which is package-private, so MiMa ignores it. Similarly, you won't find any of these APIs in the scaladocs.

In fact, I'm pretty surprised this works at all. This smells a lot to me like https://github.com/scala/bug/issues/6794.

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: If we want to get MiMa support for these, we'll have to expose them as a public `class` or `trait` somewhere, which should match the type of that `val NonEmptySet`.

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Oh, I see... Is it even possible to configure Mima to force it looking into some specific objects even if they're package private? We would need to add those 3 objects then at least...

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: You can open an issue on MiMa, but not sure what can be done there.

https://github.com/typelevel/cats/blob/708cbaba2e6ac07a5ac1819b6c18b0c83a2faaba/core/src/main/scala/cats/data/NonEmptySet.scala#L30

Instead I propose this solution: 
```scala
sealed class NonEmptySetCompanion extends NonEmptySetInstances with Newtype`
private[data] object NonEmptySetImpl extends NonEmptySetCompanion

val NonEmptySet: NonEmptySetCompanion = NonEmptySetImpl
```

I think that should get us what we want.

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Ah, but note, even if we do my proposal and publish it as 2.8.0, it won't check bincompat for us against 2.7.0 or earlier. For that we would really need changes in MiMa.

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Does Mima check against previous snapshots created out of the current `main` branch? What if I first create a PR with public sealed class as you proposed, then we merge it and then – I wonder – will Mima be checking against the already merged code, wdyt?

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: And another question... Should it be really a class? Or will a public sealed trait be enough there as well?

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: > Does Mima check against previous snapshots created out of the current main branch?

No, it's not currently configured to check snapshots, but in theory we could try and configure that. Technically, we don't guarantee binary-compatibility against snapshots, and it might be annoying if we add something but then remove it or change it before the release.

> And another question... Should it be really a class? Or will a public sealed trait be enough there as well?

Yeah, `class`, `trait`, whatever :) so long as its public.

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Btw, there is one other option. Cats has a manual bincompat test here:
https://github.com/typelevel/cats/blob/main/binCompatTest/src/main/scala-2/catsBC/MimaExceptions.scala

It actually compiles against an old Cats artifact, and then attempts to run it with the current branch. So that is the true test for bincompat (unlike MiMa, which is just doing an analysis). So in theory we could add use all of these methods in there, and it will be tested as part of CI even if MiMa can't check it for us.

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Oh, cool, I didn't know about it. So do you mean – if I add an appropriate check for that implicit there, then it will be checked against the previous published artifact and we can see if really BinCompat or not – is that correct?

Comment (armanbilge) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Yes, that's right. If you do something which uses `catsDataHashForNonEmptySet` it _should_ fail since you removed the `Order`.

Comment (DavidGregory084) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: You could possibly try a bytecode level compatibility checker like [Revapi](https://revapi.org/revapi-site/main/getting-started.html) or [SigTest](https://wiki.openjdk.java.net/display/CodeTools/SigTest) as a one off, although I don't know how well they will work on Scala projects

Comment (satorg) [reply core/src/main/scala/cats/data/NonEmptySet.scala:431]: Hopefully handled in #4219.

**Comment by satorg** [comment]
Hi @armanbilge. I realized that warnings for Scala 2.12 for unused .. well, everything .. were suppressed entirely in [sbt-typelevel](https://github.com/typelevel/sbt-typelevel/blame/v0.4.12/settings/src/main/scala/org/typelevel/sbt/TypelevelSettingsPlugin.scala#L88):
```
      val warnings212 = Seq("-Xlint:-unused,_")
```
So.. I wonder – was it made intentionally and can we assume now that the warnings regarding unused stuff will never show up for Scala 2.12?

**Comment by armanbilge** [comment]
Oh, interesting 😅 sbt-typelevel inherited those settings from sbt-spiewak which derived them from sbt-tpolecat. So probably intentional. In any case, it's what all the builds are using now, and many were using before too, so it probably works just fine.

**Comment by satorg** [comment]
That's fine... Just curious though – is it just because 2.12 is sunsetting and no one likes messing around with it? Or is it something more special...

**Comment by armanbilge** [comment]
Here's the original commit. https://github.com/djspiewak/sbt-spiewak/commit/95a45d0ea91a015c1e38e9f56d5533bb3acd8f28

**Comment by satorg** [comment]
Oh that's interesting... It was introduces even before Scala 2.13.0 was released. Seems like Daniel just didn't like that sort of warnings at all 😄

**Comment by armanbilge** [comment]
Btw, remember sbt-typelevel hosts itself and sbt uses Scala 2.12. I'll have to check later but I'm pretty sure I have unused warnings when I develop it. So maybe it's enabled by something else 🤔

**Comment by satorg** [comment]
Well, for Cats it definitely disabled because if I enable it temporarily and re-compile for 2.12, then the "unused" warnings begin coming in numbers.

**Comment by armanbilge** [comment]
You're right. I was dreaming about sbt-typelevel, seems I don't have unused warnings there at all 🙃

**Comment by armanbilge** [comment]
@satorg I've bumped sbt-typelevel to a snapshot with your PR (thanks again for doing that!). That should unblock your work, and we'll make sure to get on a stable release before we merge :)

**Comment by satorg** [comment]
@armanbilge Initially I intended to fix compiler warnings in a bulk. And I completed with "kernel" modules quite a long ago. But turned out that "core" modules have quite a lot of warnings like these:
```
[warn] /Users/.../typelevel/cats/core/src/main/scala/cats/syntax/bitraverse.scala:26:22: parameter value evidence$1 in method catsSyntaxBitraverse is never used
[warn]   implicit final def catsSyntaxBitraverse[F[_, _]: Bitraverse, A, B](fab: F[A, B]): BitraverseOps[F, A, B] =
[warn]                      ^
[warn] /Users/.../typelevel/cats/core/src/main/scala/cats/syntax/bitraverse.scala:31:22: parameter value evidence$2 in method catsSyntaxNestedBitraverse is never used
[warn]   implicit final def catsSyntaxNestedBitraverse[F[_, _]: Bitraverse, G[_], A, B](
[warn]                      ^
[warn] /Users/.../typelevel/cats/core/src/main/scala/cats/syntax/bitraverse.scala:48:22: parameter value evidence$4 in method catsSyntaxBitraverseBinCompat0 is never use
[warn]   implicit final def catsSyntaxBitraverseBinCompat0[F[_, _]: Bitraverse, A, B](
[warn]                      ^
```
which require quite gentle and precise handling I think.

But since the "kernel" modules seem to be ready, I'd like to separate the work into multiple PRs and thus keep the kernel-related part within this PR to let it being reviewed and hopefully merged. Meanwhile I can continue working on the "core" part.

Would it work, wdyt?

**Comment by satorg** [comment]
Ok nice, thank you. I think this PR is pretty much ready then.

File: kernel/src/main/scala/cats/kernel/instances/MapInstances.scala
Comment (armanbilge) [inline kernel/src/main/scala/cats/kernel/instances/MapInstances.scala:50]: I'm actually not sure about this change. Even though the `Hash[K[` instance is not used, arguably it does serve as evidence that `K` _is_ hashable (e.g., it's not a `Function`).

See Oscar's interesting take on this in https://github.com/typelevel/cats/pull/4193#issuecomment-1119933018.

Comment (satorg) [reply kernel/src/main/scala/cats/kernel/instances/MapInstances.scala:50]: Well, I'm ok to roll it back, but then another question is popping up actually: why the `MapHash` class uses the universal `.hashCode` on keys rather than `Hash[K]`? I am totally aware that Scala's `Map` itself uses universal hashes – but so what? The purpose of `MapHash` is to calculate `Hash[Map[K, V]]` out of the passed `Map` but it does not treat it as the "map", actually. All it does – just enumerates all its keys and values and produces a resulting hash value.

To me it seems that it would be better to make `MapHash` relying `Hash[K]` upon the calculation instead of the universal hash for `K`. Or are there come other considerations which I am just not aware of?

Comment (armanbilge) [reply kernel/src/main/scala/cats/kernel/instances/MapInstances.scala:50]: > but so what?

😀 yeah, ok, great point. I think it dates back to when `Hash` was introduced in https://github.com/typelevel/cats/pull/1712 and I don't see any discussion on that specifically. Probably a separate PR, but I think it makes sense to change it (anyway we currently have a law claiming that `Hash` is consistent with `hashCode` so in theory it should not change anything).

Comment (armanbilge) [reply kernel/src/main/scala/cats/kernel/instances/MapInstances.scala:50]: Ah, well, in theory you could end up with something inconsistent with `MapEq`? Since Scala forces our hand there to use universal equality.

https://github.com/typelevel/cats/blob/f8f6eff94872cba8bead3ae967d5b4f60f1b068d/kernel/src/main/scala/cats/kernel/instances/MapInstances.scala#L87-L97

But if we can't use the `Hash[K]`, I do see an argument we shouldn't ask for it. Ugh.

Comment (armanbilge) [reply kernel/src/main/scala/cats/kernel/instances/MapInstances.scala:50]: So still maybe the best strategy for now is to keep the `Hash[K]` and suppress the warning. It still has some use, by proving `K` is hashable, and it leaves the door open for us to decide what we want to do.

Anyway, we have a real `HashMap` now so we shouldn't stress over this too much ... we should figure out how to land the other PR instead.

File: kernel/src/main/scala-2.12/cats/kernel/compat/unused.scala
Comment (satorg) [reply kernel/src/main/scala-2.12/cats/kernel/compat/unused.scala:31]: I'm on a fence with this implementation of `unused` for Scala 2.12.
Turns out that it does not always work out everywhere (due to 2.12 limitations I guess). For example, in this case (let's name it "A")
https://github.com/typelevel/cats/blob/f8f6eff94872cba8bead3ae967d5b4f60f1b068d/core/src/main/scala/cats/data/Const.scala#L42-L43
it works just fine.

But on the other hand for this case "B":
https://github.com/typelevel/cats/blob/f8f6eff94872cba8bead3ae967d5b4f60f1b068d/core/src/main/scala/cats/ApplicativeError.scala#L356-L362

it produces another quite weird warning (instead of "parameter is never used"):
```
[warn] /Users/.../typelevel/cats/core/src/main/scala/cats/ApplicativeError.scala:359:7: value NT is deprecated
[warn]       try {
[warn]       ^
```
And there are a few more similar cases in `core`. I figured out that for the latter case ("B") the following implementation of `@unused` works better:
```scala
final private[cats] class unused extends scala.annotation.nowarn("msg=is never used")
```
But it does not work at all for the former case ("A") – Scala 2.12 warns the `@nowarn` annotation does not suppress any warning.

So seems that to  handle differences between 2.12 and 2.13 properly we need to have 2 different compatibility implementations for `@unused`: (1) `@deprecated`-based and (2) `@nowarn`-based. And pick one of them depending on a particular case.

@armanbilge any thoughts on this?

**Comment by armanbilge** [comment]
@satorg I'm sorry, I haven't forgotten about this PR.

The truth is I've been hesitating, because I'm weighing what I think are fairly straightforward changes here with the fact it requires putting Cats on an sbt-typelevel milestone. There are some non-trivial changes coming to sbt-typelevel and I'd rather not have to worry about accidentally breaking the Cats build (the fact that Cats Effect is stuck on the milestone is stress enough!).

- upgrade to [laika 0.19](https://github.com/typelevel/sbt-typelevel/issues/228)
- rework to use [scalac-options](https://github.com/typelevel/scalac-options)
- replacing the [sonatype client](https://github.com/davenverse/sonatype/)

In theory we could backport your scalac warning fixes to sbt-tl 0.4.x. However if you're eager to work on that what I'd actually love help with is integrating with the scalac-options library for 0.5.x.

**Comment by satorg** [comment]
@armanbilge yeah, no worries. I also feel a bit hesitant about pushing this work out. Because before continuing it would be better to stabilize the build system itself.

> However if you're eager to work on that what I'd actually love help with is integrating with the scalac-options library for 0.5.x.

Ah ok, I think I can help with that. Thanks.

**Comment by satorg** [comment]
Converted to "Draft" to put on hold until work on scalac compiler options is stabilized.

**Comment by som-snytt** [comment]
I arrived via good first issues [sic]. I know about options on Scala 2 and somewhat on Scala 3, and I generally endorse using `-Xlint`, also it's not "fatal warnings" but `-Werror` which I pronounce "worry". I'm interested in learning how cats can be supported by the options cross-compilation, but maybe you're already working out a solution.

**Comment by armanbilge** [comment]
@satorg @som-snytt if there are folks motivated to work on this, I'm thinking we should just update Cats to the sbt-typelevel milestone to unblock this work.

In https://github.com/typelevel/cats/pull/4187#issuecomment-1243184576 I was worried about too many changes happening in sbt-typelevel, but besides the Laika upgrade the other objectives have been dormant, and I don't think that they are going to land in 0.5 after all. That's okay 🙁

**Comment by satorg** [comment]
@armanbilge @som-snytt if I remember correctly, there are quite a few warnings that are handled differently in different Scala versions (i.e. 2.12, 2.13, 3). And that is what we tried to address with  [scalac-compat](https://github.com/typelevel/scalac-compat). I think, it could come handy for Cats too. Are we fine with bringing it in there?

**Comment by armanbilge** [comment]
Yes, scalac-compat has been wonderful! We are using it in Cats Effect now as well.

**Comment by satorg** [comment]
scala/scala3#20536 (for reference)

**Comment by satorg** [comment]
Superseded by #4614.

**Comment by nifledcar7-blip** [comment]
> This is an ongoing effort for suppressing compiler warnings that keep plaguing the project.
> 
> Suppresses the warnings in kernel* modules for all Scala versions.

http4s/http4s#6300

**Pull request closed without merging**
